sklearn/neighbors/binary_tree.pxi

```diff
@@ -601,11 +601,15 @@ cdef class NeighborsHeap:
             self._sort()
         return self.distances_arr, self.indices_arr
 
-    cdef inline DTYPE_t largest(self, ITYPE_t row) except -1:
+    cdef inline DTYPE_t largest(self, ITYPE_t row) nogil except -1:
         """Return the largest distance in the given row"""
         return self.distances[row, 0]
 
-    cpdef int push(self, ITYPE_t row, DTYPE_t val, ITYPE_t i_val) except -1:
+    def push(self, ITYPE_t row, DTYPE_t val, ITYPE_t i_val):
+        return self.cpush(row, val, i_val)
+
+    cdef int cpush(self, ITYPE_t row, DTYPE_t val,
```

Comment:
I would rename `cpush` to `_push`. I think this is the way we do it in other Cython files of scikit-learn.